### ISSUE TITLE
Simulator uses too much memory for larger job trace files

### ISSUE THREAD

**Issue by sampollard** [flux-framework/flux-sched#288]
Here is an example:
<img width="1269" alt="flux-memory" src="https://user-images.githubusercontent.com/6352576/34232952-a43a9fac-e597-11e7-91fa-1ef71ffe4182.png">

It appears that flux simulator is using too much memory.

Here is the command I ran:
```
./sim_sbatch.sh --plugin sched.backfill --job back_Sep_-1 --trace /home/users/spollard/llnl/data/Sep.2014.slurm.cab --rdl /home/users/spollard/llnl/data/cab.ordered.lua --log_dir logs/back_Sep_-1 --num_jobs -1 2> logs/back_Sep_-1.error.log > logs/back_Sep_-1.output.log &
```

Here is the output:
```
FLUX CMD = flux
JOB NAME = back_Sep_-1
TRACE_FILE = /home/users/spollard/llnl/data/Sep.2014.slurm.cab
SCHED_PLUGIN = sched.backfill
EXTRA FLAGS =
LOG DIR = logs/back_Sep_-1
NUM JOBS = -1
```

Here is the error:
```
./sim_sbatch.sh: line 113: 33304 Killed                  ${FLUX_CMD} start -o,-Spersist-filesystem=${PERSIST_DIR} ./initial_program ${TRACE_FILE} ${RESULTS_DIR} --sched_plugin ${SCHED_PLUGIN} --log_dir ${LOG_DIR} --rdl ${RDL} --num_jobs ${NUM_JOBS} ${PLUGIN_OPTS} ${EXTRA_FLAGS}
```

I am using
* Flux-core version: e2b5e8ddeaca5f08d2efd636f72a9e1f871ae53d
* Flux-sched version: 629631558c2884b4ba2f6ca4106a8b0bcff8fa83 (from github.com/sampollard/flux-sched)

This is performed using a modified version of @SteVwonder Python simulation script (-1 means run all jobs). My own modified version of the script is located at: https://lc.llnl.gov/bitbucket/users/pollard4/repos/flux-sim-aux/browse (this is internal to LLNL but I can provide the scripts if need be).

I won't post the broker file since it's 63GB, but here's the last few lines:
```
2017-12-14T21:30:50.137090Z sched.debug[0]: attempting job 55205 state change from selected to allocated
2017-12-14T21:30:50.143188Z sched.debug[0]: job (55205) assigned new state: runrequest
2017-12-14T21:30:50.143258Z sched.debug[0]: job 55205 runrequest
2017-12-14T21:30:50.143298Z sched.debug[0]: scheduler timer: events + loop took 0.007603 seconds
2017-12-14T21:30:50.143381Z sched.debug[0]: sent a reply request: {"sim_time": 1413458638.0202873, "mod_name": "sched", "event_timers": {"sched": -1.0, "sim_exec": 1413458638.0203872, "submit": -1.0}}
2017-12-14T21:30:50.143583Z sched.debug[0]: JscEvent being queued - JSON: {"jobid": 55205, "state-pair": {"ostate": 7, "nstate": 8}}, errnum: 0
2017-12-14T21:30:50.143937Z sim.debug[0]: Triggering sim_exec.  Curr sim time: 1413458638.020387
2017-12-14T21:30:50.144468Z sim_exec.debug[0]: received a request (sim_exec.run.55205)
2017-12-14T21:30:50.144893Z sim_exec.debug[0]: queued the running of jobid 55205
2017-12-14T21:30:50.144992Z sim_exec.debug[0]: received a trigger (sim_exec.trigger: {"sim_time": 1413458638.0203872, "event_timers": {"sched": -1.0, "sim_exec": -1.0, "submit": -1.0}}
2017-12-14T21:30:50.157091Z sim_exec.debug[0]: job (55205) assigned new state: starting
2017-12-14T21:30:50.157516Z sched.debug[0]: JscEvent being queued - JSON: {"jobid": 55205, "state-pair": {"ostate": 8, "nstate": 9}}, errnum: 0
2017-12-14T21:30:50.165842Z sim_exec.debug[0]: job (55205) assigned new state: running
2017-12-14T21:30:50.166243Z sched.debug[0]: JscEvent being queued - JSON: {"jobid": 55205, "state-pair": {"ostate": 9, "nstate": 11}}, errnum: 0
2017-12-14T21:30:50.169837Z sim_exec.info[0]: job 55205's state to starting then running
2017-12-14T21:30:50.171227Z sim_exec.debug[0]: next sim_exec event set to 1413458648.026263
2017-12-14T21:30:50.171334Z sim_exec.debug[0]: sent a reply request: {"sim_time": 1413458638.0203872, "mod_name": "sim_exec", "event_timers": {"sched": -1.0, "sim_exec": 1413458648.0262635, "submit": -1.0}}
2017-12-14T21:30:50.171605Z sim.debug[0]: Triggering sim_exec.  Curr sim time: 1413458648.026263
2017-12-14T21:30:50.171971Z sim_exec.debug[0]: received a trigger (sim_exec.trigger: {"sim_time": 1413458648.0262635, "event_timers": {"sched": -1.0, "sim_exec": -1.0, "submit": -1.0}}
2017-12-14T21:30:50.172063Z sim_exec.info[0]: Job 55203 completed
2017-12-14T21:30:50.185446Z broker.err[0]: Run level 2 Killed (rc=137) 175150.0s
2017-12-14T21:30:50.224977Z broker.info[0]: Run level 3 starting
2017-12-14T21:31:07.335344Z broker.info[0]: shutdown in 1.000s: run level 2 Killed
```

Furthermore, the data in the persist directory (sqlite file) is 11G.

Please let me know if there is anything else I can provide.

**Comment by dongahn** [comment]
Do you have any idea what part of the simulator logic uses excessive amounts of memory? 

I think the size of persistent file system is known as we put our content store there. I've been using a temporary file system so that I don't have to hit my home directory quota.

**Comment by sampollard** [comment]
I don't know, but I do know there is the `zlist_t completion_times` list that the scheduler keeps. I believe for a simulation run there is never a step where this list is shrunk. Even though, that shouldn't amount to 190G.

I forgot to mention: the job trace file I am using has about 100,000 jobs. If you are a member of perfml you can find it at /usr/gapps/perfml/job-queues/cab/Sep.2014.slurm.cab

**Comment by dongahn** [comment]
Is that really 190GB? Are you measuring resident set size or virtual memory size?

**Comment by sampollard** [comment]
Ah, 190GB is the virtual memory size. That screenshot I posted says 120G resident set size.

**Comment by dongahn** [comment]
120GB still big. What system did you use and how much physical memory do you have on it?

**Comment by sampollard** [comment]
128GB of memory:
```
$ uname -a
Linux sansa 4.4.0-72-generic #93-Ubuntu SMP Fri Mar 31 14:07:41 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```
```
$ sudo lshw -short
H/W path                Device      Class      Description
==========================================================
                                    system     PowerEdge R810
/0                                  bus        0TT6JF
/0/0                                memory     64KiB BIOS
/0/400                              processor  Intel(R) Xeon(R) CPU           X7560  @ 2.27GHz
/0/400/700                          memory     512KiB L1 cache
/0/400/701                          memory     2MiB L2 cache
/0/400/702                          memory     24MiB L3 cache
/0/401                              processor  Intel(R) Xeon(R) CPU           X7560  @ 2.27GHz
/0/401/703                          memory     512KiB L1 cache
/0/401/704                          memory     2MiB L2 cache
/0/401/705                          memory     24MiB L3 cache
/0/1000                             memory     128GiB System Memory
/0/1000/0                           memory     8GiB DIMM DDR3 Synchronous 1333 MHz (0.8 ns)
...
```

**Comment by SteVwonder** [comment]
I think the best chance for tracking this down is to run the simulator under either massif or totalview's memory profiler.  I am willing to do this, but I won't be able to commit to working on it until after January.

**Comment by grondo** [comment]
closed by #440
